src/Compilers/CSharp/Portable/BoundTree/Expression.cs

```diff
@@ -0,0 +1,1447 @@
+        private static System.Runtime.CompilerServices.ConditionalWeakTable<BoundExpression, IArgument> ArgumentMappings = new System.Runtime.CompilerServices.ConditionalWeakTable<BoundExpression, IArgument>();
+
+        private static IArgument DeriveArgument(int parameterIndex, int argumentIndex, ImmutableArray<BoundExpression> boundArguments, ImmutableArray<string> argumentNames, ImmutableArray<RefKind> argumentRefKinds, ImmutableArray<Symbols.ParameterSymbol> parameters)
+        {
+            if (argumentIndex >= boundArguments.Length)
+            {
+                // Check for an omitted argument that becomes an empty params array.
+                if (parameters.Length > 0 && parameters[parameters.Length - 1].IsParams)
+                {
+                    return new Argument(ArgumentKind.ParamArray, parameters[parameters.Length - 1], CreateParamArray(parameters[parameters.Length - 1], boundArguments, argumentIndex));
+                }
+
+                // There is no supplied argument and there is no params parameter. Any action is suspect at this point.
+                return null;
+            }
+
+            return ArgumentMappings.GetValue(
+                boundArguments[argumentIndex],
+                (argument) =>
+                {
+                    string name = !argumentNames.IsDefaultOrEmpty ? argumentNames[argumentIndex] : null;
+                    RefKind refMode = !argumentRefKinds.IsDefaultOrEmpty ? argumentRefKinds[argumentIndex] : RefKind.None;
+
+                    if (name == null)
+                    {
+                        return
+                            refMode == RefKind.None
+                            ? ((argumentIndex >= parameters.Length - 1 && parameters.Length > 0 && parameters[parameters.Length - 1].IsParams)
+                                ? (IArgument)new Argument(ArgumentKind.ParamArray, parameters[parameters.Length - 1], CreateParamArray(parameters[parameters.Length - 1], boundArguments, argumentIndex))
+                                : new SimpleArgument(parameters[parameterIndex], argument))
+                            : (IArgument)new Argument(ArgumentKind.Positional, parameters[parameterIndex], argument);
+                    }
+
+                    return new Argument(ArgumentKind.Named, parameters[parameterIndex], argument);
+                });
+        }
+        
+        private static IExpression CreateParamArray(IParameterSymbol parameter, ImmutableArray<BoundExpression> boundArguments, int firstArgumentElementIndex)
+        {
+            if (parameter.Type.TypeKind == TypeKind.Array)
+            {
+                IArrayTypeSymbol arrayType = (IArrayTypeSymbol)parameter.Type;
+                List<IExpression> paramArrayArguments = new List<IExpression>();
+                for (int index = firstArgumentElementIndex; index < boundArguments.Length; index++)
+                {
+                    paramArrayArguments.Add(boundArguments[index]);
+                }
```

Comment:
Consider using ImmutableArray&lt;IExpression&gt; for paramArrayArguments. And perhaps shorten to: var paramArrayArguments = ImmutableArray.CreateRange(boundArguments.Skip(firstArgIndex));